### PR TITLE
Example Routes

### PR DESCRIPTION
**Pull request by ghost**
I have written an example route to show a suggestion for how we could fit the system together and standarize how we write routes. 

I have used AJV for validating I/O, using that to type cast, so we validating all data in the system, and then connected it to the db and the logging system. 

This should be good enough to discuss possible solutions for a template for new routes. 

I think the most "controversial" part i added to the code now would be the formater for SQL code. It merges the two syntaxes for interacting with a PSQL server. We have native pg-node, and pg-format. pg-format is more powerful but not always necessary, and they do not share a syntax. The function takes in both syntaxes and returns a formated SQL statement. 

 The function in question is from Line 36 in databasefunctions.

### CODE REVIEW

File: backend/src/api/routes/users.ts [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+const app: express.Application = express();
+
+const newCompany = async (req: express.Request, res: express.Response, next: NextFunction): Promise<void | never> => {
+  try {
+    const values = (await extractPayload(req)) as Fields;
+    if (instanceOfCompany(values)) {
+      const { email, orgnum, username, salt, passhash, name }: Company = values;
+      logger.info(`Creating new company ${username}`);
+      const pool: Pool = await genereateAdminConnection();
+      const poolClient: PoolClient = await getPoolClient(pool);
```
Comment (mariusaarsnes) [inline backend/src/api/routes/users.ts:21]: What is the purpose of these lines, is it to assert that the user is an admin?

File: backend/src/api/routes/users.ts [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+  res: express.Response,
+  next: NextFunction,
+): Promise<void | never> => {
+  try {
+    const values = (await extractPayload(req)) as Fields;
+    if (instanceOfCompanyUser(values)) {
+      const { email, username, salt, passhash, givenname, surename, company }: CompanyUser = values;
+      logger.info(`Creating new companyuser ${username}`);
+      const pool: Pool = await genereateAdminConnection();
+      const poolClient: PoolClient = await getPoolClient(pool);
```
Comment (mariusaarsnes) [inline backend/src/api/routes/users.ts:44]: same as above

File: backend/src/utils/interfaces.ts [hunk 3]
```diff
@@ -0,0 +1,25 @@
+//TODO find a way to enforce regex format for email inside an interface.
+export interface Company extends Object {
```
Comment (mariusaarsnes) [inline backend/src/utils/interfaces.ts:1]: pretty sure i have talked about this before, but i really have problems understanding how/why a company should be a user. 
Who should have access to this user and what should this user be able to do that a company user should not?

File: backend/src/utils/interfaces.ts [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+  orgnum: bigint;
+  username: string;
+  salt: string;
+  passhash: string;
+  name: string;
+}
+
+export interface CompanyUser extends Object {
+  email: string;
+  telephone?: bigint;
```
Comment (mariusaarsnes) [inline backend/src/utils/interfaces.ts]: might as well require it

File: backend/src/utils/interfaces.ts [hunk 5]
```diff
@@ -0,0 +1,25 @@
+//TODO find a way to enforce regex format for email inside an interface.
```
Comment (mariusaarsnes) [inline backend/src/utils/interfaces.ts]: Should this comment be removed?

**Review by mariusaarsnes — CHANGES_REQUESTED** [review]

File: backend/src/api/routes/users.ts (hunk 1 above)
Comment (ghost) [reply backend/src/api/routes/users.ts:21]: Those lines get a connection to the database that are admins. 

It is the only way to do "Create role" and "grant" statements.

File: backend/src/utils/interfaces.ts (hunk 3 above)
Comment (ghost) [reply backend/src/utils/interfaces.ts:1]: It was mostly to control the users, and their permission, but i just realised what i think you've been trying to convey the whole time... 😅 

I'll remove the company as a user from the db and their role creation from the route. 

I'll still neeed the company interface, that interface is only used to ensure that when a new company is created that we know what type the dataobjects has.

File: backend/src/utils/interfaces.ts (hunk 4 above)
Comment (ghost) [reply backend/src/utils/interfaces.ts]: I can update it to be required for both the db and the interface.

File: backend/src/utils/interfaces.ts (hunk 5 above)
Comment (ghost) [reply backend/src/utils/interfaces.ts]: I guess yes, i got the enforment through AJV, so it should be fine, as long as we validate everything.

File: backend/src/utils/interfaces.ts (hunk 3 above)
Comment (mariusaarsnes) [reply backend/src/utils/interfaces.ts:1]: yes, nice. Then we are on the same page :+1:

File: backend/src/api/routes/users.ts (hunk 1 above)
Comment (mariusaarsnes) [reply backend/src/api/routes/users.ts:21]: Okay, so they are not used to actually establish that the user has admin rights?

Is there some middleware restricting access to routes which require admin privileges?

**Review by mariusaarsnes — COMMENTED** [review]
I have a follow-up question for one of my previous comments. If the answer is yes, then just resolve the conversation.

File: backend/src/api/routes/users.ts (hunk 1 above)
Comment (ghost) [reply backend/src/api/routes/users.ts:21]: The next commit i'm adding is the one containing the rights management for express.

So the next PR will include OICD auth using OW4. 

So the short answer is no. These user do not have admin rights, they only need an admin on the db to be created as roles in the db.

**Review by mariusaarsnes — APPROVED** [review]

**Pull request merged**